pricegraph/src/orderbook.rs

```diff
@@ -149,12 +162,88 @@ impl Orderbook {
+    ///
+    /// Note that currently, user buy token balances are not incremented as a
+    /// result of filling orders along a path.
+    pub fn fill_path(&mut self, path: &[TokenId]) -> Option<f64> {
+        let capacity = self.find_maximum_capacity(path)?;
+        self.fill_path_with_capacity(path, capacity)
+            .expect("failed to fill detected path with capacity");
+
+        Some(capacity)
+    }
+
+    /// Gets the maximum capacity of a path expressed in an amount of the
+    /// starting token. Returns `None` if the path doesn't exist.
+    fn find_maximum_capacity(&self, path: &[TokenId]) -> Option<f64> {
+        let mut capacity = f64::INFINITY;
+        let mut transient_price = 1.0;
+        for pair in pairs_on_path(path) {
+            let order = self.orders.best_order_for_pair(pair)?;
+            transient_price *= order.price;
+
+            let sell_amount = order.get_effective_amount(&self.users);
+            capacity = num::min(capacity, sell_amount * transient_price);
+        }
+
+        Some(capacity)
+    }
+
+    /// Pushes flow through a path of orders reducing order amounts and user
+    /// balances as well as updating the projection graph by updating the
+    /// weights to reflect the new graph.
+    ///
+    /// Note that currently, user buy token balances are not incremented as a
+    /// result of filling orders along a path.
+    fn fill_path_with_capacity(
+        &mut self,
+        path: &[TokenId],
+        capacity: f64,
+    ) -> Result<(), IncompletePathError> {
+        let mut transient_price = 1.0;
+        for pair in pairs_on_path(path) {
+            let (order, user) = self
+                .best_order_with_user_for_pair_mut(pair)
+                .ok_or_else(|| IncompletePathError(pair))?;
+
+            transient_price *= order.price;
+
+            // NOTE: `capacity` here is a buy amount, so we need to divide by
+            // the price to get the sell amount being filled.
+            let fill_amount = capacity / transient_price;
+
+            order.amount -= fill_amount;
+            let new_balance = user.deduct_from_balance(order.pair.sell, fill_amount);
+
+            if new_balance.is_none() {
+                self.update_projection_graph_node(pair.sell);
+            } else if order.amount <= 0.0 {
+                self.update_projection_graph_edge(pair);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Gets a mutable reference to the cheapest order for a given token pair
+    /// along with the user that placed the order. Returns `None` if there are
+    /// no orders for that token pair.
+    fn best_order_with_user_for_pair_mut(
+        &mut self,
+        pair: TokenPair,
+    ) -> Option<(&'_ mut Order, &'_ mut User)> {
+        let Self { orders, users, .. } = self;
+
+        let order = orders.best_order_for_pair_mut(pair)?;
+        let user = users
+            .get_mut(&order.user)
+            .expect("missing user for existing order");
```

Comment:
nit: Would it make debugging easier if we included the user/order pair here?

Reply:
Good point! although I REALLY hope this assertion never fails :P